Add initial support for PostgreSQL

dqminh opened the pull request:
Closes #600 

As title stated :)

A quick run with `docker-compose -f docker-compose.postgresql.yml up` and testing with a local docker client pushing to local registry doesn't show any errors so hopefully i didn't miss anything.

cc @thaJeztah @endophage

thaJeztah commented on migrations/migrate.sh:
```diff
@@ -7,45 +7,49 @@ iter=0
 				exit 1;
 			fi
-			echo "waiting for notarymysql to come up."
+      echo "waiting for $DB_URL to come up."
```
Looks like you're using spaces here (and a couple of other places), should be tabs in shell scripts :smile:

dqminh replied:
there goes my secret plan to convert tabs to spaces.

thaJeztah commented:
Thanks! I'll leave code-review to the notary team, but I like PostgreSQL, so 😍

riyazdf commented:
@dqminh thank you for adding this feature!  Could you please rebase?  I think you're hitting a linter CI issue that we addressed a few days ago.

We should have a testing infrastructure for postgres if we choose to support it - I think a good starting point would be adding a case to `buildscripts/integrationtest.sh` if you don't mind :)

dqminh commented:
@riyazdf i cant get the integration test to pass in my local environment. Is there any special tricks ?

Here is part of the build failure with rethinkdb as an example:

```
[31msigner_1     |[0m successfully reached and updated RethinkDB
[31msigner_1     |[0m {"level":"info","msg":"Version: 0.3, Git commit: 2a49e9a","time":"2016-08-17T11:10:27Z"}
[31msigner_1     |[0m {"level":"debug","msg":"Trusting 2 certs","time":"2016-08-17T11:10:27Z"}
[31msigner_1     |[0m {"level":"debug","msg":"Default Alias: timestamp_1","time":"2016-08-17T11:10:27Z"}
[31msigner_1     |[0m {"level":"debug","msg":"attempting to connect user signer to host rdb-proxy.rdb","time":"2016-08-17T11:10:27Z"}
[31msigner_1     |[0m {"level":"debug","msg":"Trusting 1 certs","time":"2016-08-17T11:10:27Z"}
[34mserver_1     |[0m trying to contact RethinkDB for 30 seconds before failing
[34mserver_1     |[0m {"level":"info","msg":"Version: 0.3, Git commit: 2a49e9a","time":"2016-08-17T11:10:26Z"}
[34mserver_1     |[0m {"level":"debug","msg":"Trusting 1 certs","time":"2016-08-17T11:10:26Z"}
[34mserver_1     |[0m {"level":"info","msg":"Using remote signing service","time":"2016-08-17T11:10:26Z"}
[34mserver_1     |[0m {"level":"info","msg":"Using rethinkdb backend","time":"2016-08-17T11:10:26Z"}
[34mserver_1     |[0m {"level":"debug","msg":"attempting to connect admin to host rdb-proxy.rdb","time":"2016-08-17T11:10:26Z"}
[34mserver_1     |[0m {"level":"debug","msg":"Trusting 1 certs","time":"2016-08-17T11:10:26Z"}
[36;1mclient_1     |[0m make: Entering directory '/go/src/github.com/docker/notary'
[36;1mclient_1     |[0m pkcs11 import was not found anywhere without a build tag, yay
[36;1mclient_1     |[0m + /go/src/github.com/docker/notary/bin/notary
[34mserver_1     |[0m {"level":"debug","msg":"creating user server for db notaryserver","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m successfully reached and updated RethinkDB
[34mserver_1     |[0m {"level":"info","msg":"Version: 0.3, Git commit: 2a49e9a","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m {"level":"debug","msg":"Trusting 1 certs","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m {"level":"info","msg":"Using remote signing service","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m {"level":"info","msg":"Using rethinkdb backend","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m {"level":"debug","msg":"attempting to connect user server to host rdb-proxy.rdb","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m {"level":"debug","msg":"Trusting 1 certs","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m {"level":"info","msg":"Starting Server","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m {"level":"info","msg":"Enabling TLS","time":"2016-08-17T11:10:29Z"}
[34mserver_1     |[0m {"level":"info","msg":"Starting on :4443","time":"2016-08-17T11:10:29Z"}
[36;1mclient_1     |[0m + client
[36;1mclient_1     |[0m make: Leaving directory '/go/src/github.com/docker/notary'
[34mserver_1     |[0m {"go.version":"go1.6.3","http.request.host":"notary-server:4443","http.request.id":"695d2c85-9e5b-43e1-bb0b-05a6f52e3b55","http.request.method":"GET","http.request.remoteaddr":"172.28.0.8:54948","http.request.uri":"/","http.request.useragent":"curl/7.38.0","level":"info","msg":"metadata not found: You have requested metadata that does not exist.","time":"2016-08-17T11:10:39Z"}
[36;1mclient_1     |[0m {"errors":[{"code":"METADATA_NOT_FOUND","message":"You have requested metadata that does not exist."}]}
[34mserver_1     |[0m {"go.version":"go1.6.3","http.request.host":"notary-server:4443","http.request.id":"695d2c85-9e5b-43e1-bb0b-05a6f52e3b55","http.request.method":"GET","http.request.remoteaddr":"172.28.0.8:54948","http.request.uri":"/","http.request.useragent":"curl/7.38.0","http.response.contenttype":"application/json; charset=utf-8","http.response.duration":"442.864µs","http.response.status":404,"http.response.written":104,"level":"info","msg":"response completed","time":"2016-08-17T11:10:39Z"}
[36;1mclient_1     |[0m Traceback (most recent call last):
[36;1mclient_1     |[0m   File "buildscripts/testclient.py", line 371, in <module>
[36;1mclient_1     |[0m     run()
[36;1mclient_1     |[0m   File "buildscripts/testclient.py", line 361, in run
[36;1mclient_1     |[0m building a new client binary
[36;1mclient_1     |[0m ---
[36;1mclient_1     |[0m ---- Initializing a repo, adding a target, and pushing ----
[36;1mclient_1     |[0m 
[36;1mclient_1     |[0m $ /go/src/github.com/docker/notary/bin/notary -c cmd/notary/config.json -d /tmp/tmpnPDD4Q_main init 809bb49b202f401a90735179ad4345cf
[36;1mclient_1     |[0m No root keys found. Generating a new root key...
[36;1mclient_1     |[0m 
[36;1mclient_1     |[0m * fatal: client is offline
[36;1mclient_1     |[0m 
[36;1mclient_1     |[0m     Tester(repo_name, Client(server, username_passwd)).run()
[36;1mclient_1     |[0m   File "buildscripts/testclient.py", line 302, in run
[36;1mclient_1     |[0m     test_func(tempfile, tempdir)
[36;1mclient_1     |[0m   File "buildscripts/testclient.py", line 157, in basic_repo_test
[36;1mclient_1     |[0m     self.client.run(["init", self.repo_name], self.dir)
[36;1mclient_1     |[0m   File "buildscripts/testclient.py", line 140, in run
[36;1mclient_1     |[0m     raise CalledProcessError(retcode, command, output=output)
[36;1mclient_1     |[0m subprocess.CalledProcessError: Command '['/go/src/github.com/docker/notary/bin/notary', '-c', 'cmd/notary/config.json', '-d', '/tmp/tmpnPDD4Q_main', 'init', '809bb49b202f401a90735179ad4345cf']' returned non-zero exit status 1
[36;1mintegrationrethink_client_1 exited with code 1
[0mAborting on container exit...
Makefile:130: recipe for target 'integration' failed
```

For mysql, it's either this or it timed out when trying to reach the server.

cyli commented:
@dqminh This is just a stab in the dark... It looks like the signer connected ok with the DB but the server failed to do so immediately after, so could you increase the connection timeout here:  https://github.com/docker/notary/blob/master/migrations/migrate.sh#L14 and here: https://github.com/docker/notary/blob/master/migrations/migrate.sh#L39 to maybe 60, and see if that helps at all?

cyli commented:
Actually ignore my comment.  :(

cyli commented:
Could you instead change:  https://github.com/docker/notary/blob/master/buildscripts/testclient.py#L100 to say

```
self.client = [binary, "-D", "-c", "cmd/notary/config.json"]
```

This should hopefully print out debug output for the client so we can see why it can't connect with the server, although the client debug logs and the actual client output may be slightly out of sync.

dqminh commented:
```
client_1     | time="2016-08-17T21:22:13Z" level=error msg="could not reach https://notary-server:4443: Get https://notary-server:4443/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"
client_1     | time="2016-08-17T21:22:13Z" level=info msg="continuing in offline mode"
```

So it looks like something is probably wrong with DNS inside my VM. Amazingly enough when i tried it with Docker for Mac it works 😵 
Going to continue with the thing that works until i have more time to figure out the DNS situation :/

riyazdf commented on signer/keydbstore/postgresql_test.go:
```diff
@@ -0,0 +1,57 @@
+)
+
+func init() {
+	// Get the MYSQL connection string from an environment variable
```
nit typo: "PostgreSQL"

riyazdf commented on utils/configuration_test.go:
```diff
@@ -159,7 +159,6 @@ func TestParseBugsnagWithEnvironmentVariables(t *testing.T) {
 // If the storage backend is invalid or not provided, an error is returned.
 func TestParseInvalidStorageBackend(t *testing.T) {
 	invalids := []string{
-		`{"storage": {"backend": "postgres", "db_url": "1234"}}`,
```
nit: can we keep this test case but with some other invalid backend name now that postgres will be valid?

dqminh commented on signer/keydbstore/sql_keydbstore_postgres.go:
```diff
@@ -0,0 +1,24 @@
+// +build postgresqldb
```
this is pretty ugly and is just here because of the tests. I would suggest that we have some mechanism to migrate the database outside of what GORM support (i.e. remove those sql type tags)

cyli replied:
Is the model should be the same no matter what type of GORM DB it is, right?  (it might not be - Postgres may have different type declarations, but right now I don't see any difference between the two files) If so, should it then remain in `sql_keydbstore.go`?  The intention was for that to be the common code for all GORM related stuff.

I think maybe we can then convert all the `mysqldb` buildtags to just `sqldb` build tags, because the logic is the same (poll DB until we can contact it) and use it for both postgres and mysql - we just pass different DB url values to it.

cyli commented:
jenkins, test this please

cyli commented on fixtures/server-config.postgres.json:
```diff
@@ -0,0 +1,23 @@
+	},
+	"storage": {
+		"backend": "postgres",
+		"db_url": "postgres://server@serverdb:5432/notaryserver?sslmode=disable"
```
Non-blocking:  if you know of a good/easy way to add SSL certs to the postgres server, that would be awesome, but from reading the docs it looks like we'd probably have to set the `PGDATA` directory to a local volume, and I'm not sure whether that's a good idea.

cyli commented:
LGTM!  Thank you adding this!

endophage commented on development.postgresql.yml:
```diff
@@ -0,0 +1,72 @@
+      DB_URL: postgres://signer@signerdb:5432/notarysigner?sslmode=disable
+    depends_on:
+      - signerdb
+  serverdb:
```
Would be good for consistency to handle these the same way we do with MySQL, a single DB container with multiple databases.

dqminh replied:
i would argue that this setup is actually better in term of separation. After all the official docs says that signer and server db should be separated :p

endophage replied:
If we can agree these compose files are for dev and testing, which they are, then having a single database server instance makes it much more convenient for us to inspect state and even `docker commit` a single running database container to preserve specific states we may want to share for debugging/demonstration purposes.

dqminh replied:
using single image also means that we need to bootstrap two different database with extra scripts similar to `notarymysql`. But the CHANGELOG also stated that:

```
+ Update `docker-compose` configuration to use official mariadb image
    + deprecate `notarymysql`
    + default to using a volume for `data` directory
    + use separate databases for `notary-server` and `notary-signer` with separate users
```

I can add extra bootstrap scripts, but i just think its cleaner this way. I also dont think commiting two databases are much harder than one.

But IANAM, Let me know what you think.

endophage replied:
What we want as maintainers is a consistent way to do things, and in this instance, you have created something that differs from the most similar existing example in the repo, mariadb. 

You should have no reason to create a `notarypostgres` image, and we would ask you to change it if you did. We deprecated `notarymysql` because there was no need for us to use a custom built image when we could simply use the official images and their built in bootstrapping, speeding up our build and test times. Postgres has this same built in boostrapping as mariadb, you can find the instructions under ["How to extend this image"](https://hub.docker.com/_/postgres/), and an example of our basic setup for mariadb, which should be very similar, [here](https://github.com/docker/notary/tree/master/notarymysql/docker-entrypoint-initdb.d).

dqminh replied:
updated. PTAL.

endophage commented on docker-compose.postgresql.yml:
```diff
@@ -0,0 +1,74 @@
+      DB_URL: postgres://signer@signerdb:5432/notarysigner?sslmode=disable
+    depends_on:
+      - signerdb
+  serverdb:
```
Same comment on consistency with MySQL setup.

endophage commented:
Thank you for updating the postgres database setup. I realize I probably seem like I'm being unreasonably pedantic, but as a maintainer, we are responsible for all the code in the repo, including that contributed by external contributors who may not be available in the future to make changes. As such, we have to make conscious efforts to reduce variation, which is unnecessary complexity, so we can mentally stay on top of all the necessary complexity.

We're taking a look at why the ineffassign checks are suddenly failing. I have some time already scheduled tomorrow afternoon to do some related style housekeeping so I'll open a PR to fix those assuming we don't find some other reason ineffassign started raising errors.

riyazdf commented:
@dqminh thanks for your work on this!

Could you please rebase? I think that should handle the unrelated CI issues :)

HuKeping commented on cmd/notary-server/config.go:
```diff
@@ -90,7 +91,7 @@ func getStore(configuration *viper.Viper, hRegister healthRegister, doBootstrap
 	case notary.MemoryBackend:
 		return storage.NewMemStorage(), nil
-	case notary.MySQLBackend, notary.SQLiteBackend:
+	case notary.MySQLBackend, notary.SQLiteBackend, notary.PostgresBackend:
```
Hi @dqminh would you mind also update the relative [configuration docs](https://github.com/docker/notary/blob/master/docs/reference/server-config.md#storage-section-required) about this backend, it's OK to do it in an separated PR, but we may forget to update it.

HuKeping commented on cmd/notary-signer/config.go:
```diff
@@ -68,7 +68,7 @@ func parseSignerConfig(configFilePath string, doBootstrap bool) (signer.Config,
 
 	// setup the cryptoservices
-	cryptoServices, err := setUpCryptoservices(config, []string{notary.MySQLBackend, notary.MemoryBackend, notary.RethinkDBBackend}, doBootstrap)
+	cryptoServices, err := setUpCryptoservices(config, []string{notary.MySQLBackend, notary.MemoryBackend, notary.PostgresBackend, notary.RethinkDBBackend}, doBootstrap)
```
I wonder if we could introduce a constant to indicate those supported database backends in the [const file](https://github.com/docker/notary/blob/master/const.go), so that we could stop keeping inject new items into the arguments list if we support more backends in the future.

maybe something like
```
var NotarySupportedDB = []string{
	MemoryBackend,
	MySQLBackend,
	SQLiteBackend,
	RethinkDBBackend,
	PostgresBackend,
}

```

HuKeping commented on server/storage/postgresql_test.go:
```diff
@@ -0,0 +1,54 @@
+		if i == 29 {
+			logrus.Fatalf("Unable to connect to %s after 60 seconds", dburl)
+		}
+		time.Sleep(2 * time.Second)
```
nit: actually, this line should be put before the `if`, otherwise when it prints out `after 60 seconds` is just only `58` seconds :)

HuKeping commented on server/storage/postgresql_test.go:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	for i := 0; i < 30; i++ {
+		gormDB, err := gorm.Open("postgres", dburl)
```
nit: could we use `notary.PostgresBackend` to replace the literal string since you have already created that one.

HuKeping commented on server/storage/postgresql_test.go:
```diff
@@ -0,0 +1,54 @@
+
+	sqldbSetup = func(t *testing.T) (*SQLStorage, func()) {
+		var cleanup1 = func() {
+			gormDB, err := gorm.Open("postgres", dburl)
```
nit: could we use notary.PostgresBackend to replace the literal string since you have already created that one.

HuKeping commented on signer/keydbstore/postgresql_test.go:
```diff
@@ -0,0 +1,57 @@
+		if i == 29 {
+			logrus.Fatalf("Unable to connect to %s after 60 seconds", dburl)
+		}
+		time.Sleep(2 * time.Second)
```
same here

HuKeping commented on signer/keydbstore/postgresql_test.go:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	for i := 0; i < 30; i++ {
+		gormDB, err := gorm.Open("postgres", dburl)
```
same here

HuKeping commented on signer/keydbstore/postgresql_test.go:
```diff
@@ -0,0 +1,57 @@
+
+	sqldbSetup = func(t *testing.T) (*SQLKeyDBStore, func()) {
+		var cleanup1 = func() {
+			gormDB, err := gorm.Open("postgres", dburl)
```
same here

HuKeping commented on signer/keydbstore/postgresql_test.go:
```diff
@@ -0,0 +1,57 @@
+		cleanup1()
+		dbStore := SetupSQLDB(t, "postgres", dburl)
+
+		require.Equal(t, "postgres", dbStore.Name())
```
same here

HuKeping commented on signer/keydbstore/postgresql_test.go:
```diff
@@ -0,0 +1,57 @@
+			gormDB.DropTable(&GormPrivateKey{})
+		}
+		cleanup1()
+		dbStore := SetupSQLDB(t, "postgres", dburl)
```
same here

HuKeping commented:
Thanks for your contribution @dqminh ! It seems need a rebase here :)

endophage commented:
@dqminh sorry we're taking a while to merge this. If you want to enable the new GH feature that lets maintainers modify your branch we'll happily do the last rebase and work out any failures so we can get it merged.

HuKeping commented:
FYI, https://github.com/blog/2247-improving-collaboration-with-forks

dqminh commented:
@endophage @HuKeping i already did that :p not sure if how its reflected on your side though.

endophage commented:
Hmmm, yeah, not seeing anything on my end. I guess I have to just try and push and see if it works.

HuKeping commented:
LGTM, pending CI

HuKeping commented:
LGTM, pending CI

cyli commented:
Thank you for following up on this @HuKeping!  Thanks for your patience @dqminh!

endophage commented on server/storage/mysql_test.go:
```diff
@@ -22,15 +22,15 @@ func init() {
 			}
 		}
-		if i == 29 {
+		if i == 30 {
```
This was already correctly retrying 30 times. The updated version will attempt 31 times. The iteration is zero indexed so when you reach line 33, `gorm.Open` at line 26 will have been called `i + 1` times for any given iteration. Therefore, when `i == 29`, `gorm.Open` has been attempted 30 times.

HuKeping replied:
@endophage you are right about the iteration, no matter using `30` or `29`, I think it is used for line 34 to log out how much time has been past.

Because of the `sleep statement`  is at the end of this `for` loop, so when `i == 30` and the process reaches line 34 to log out sth, we may have just sleep 30 times, thus 60 seconds.

endophage replied:
Yeah, I'm not overly bothered which it is in this case, just wanted to be clear the commit message isn't correct and it's a somewhat superfluous commit.

HuKeping replied:
yep,  the commit message"properly retry 30 times in mysql|postgresql_test" is not correct.

endophage commented:
LGTM

HuKeping commented:
Thank you for your work on this @dqminh !
